File: Ashutosh-Singh/PrefixAndSuffix.cpp/Matrix-Block-Sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    void calculatePrefixSum(vector<vector<int>>& psum) {
        int n = psum.size();
        int m = psum[0].size();

        // Row-wise prefix sum
        for (int i = 0; i < n; i++) {
            for (int j = 1; j < m; j++) {
                psum[i][j] += psum[i][j - 1];
            }
        }

        // Column-wise prefix sum
        for (int i = 0; i < m; i++) {
            for (int j = 1; j < n; j++) {
                psum[j][i] += psum[j - 1][i];
            }
        }
    }

    int sumRegion(vector<vector<int>>& psum, int row1, int col1, int row2, int col2) {
        int ans = psum[row2][col2];
        if (row1 > 0) ans -= psum[row1 - 1][col2];
        if (col1 > 0) ans -= psum[row2][col1 - 1];
        if (row1 > 0 && col1 > 0) ans += psum[row1 - 1][col1 - 1];
        return ans;
    }

    vector<vector<int>> matrixBlockSum(vector<vector<int>>& mat, int k) {
        vector<vector<int>> psum = mat;
        calculatePrefixSum(psum);

        for (int i = 0; i < mat.size(); i++) {
            for (int j = 0; j < mat[0].size(); j++) {
                int r1 = max(0, i - k);
                int c1 = max(0, j - k);
                int r2 = min((int)mat.size() - 1, i + k);
                int c2 = min((int)mat[0].size() - 1, j + k);
                mat[i][j] = sumRegion(psum, r1, c1, r2, c2);
            }
        }
        return mat;
    }
};

int main() {
    Solution sol;

    int rows, cols, k;
    cout << "Enter number of rows: ";
    cin >> rows;
    cout << "Enter number of columns: ";
    cin >> cols;

    vector<vector<int>> mat(rows, vector<int>(cols));

    cout << "Enter the elements of the matrix:" << endl;
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            cin >> mat[i][j];
        }
    }

    cout << "Enter the value of k: ";
    cin >> k;

    // Calling the matrixBlockSum function
    vector<vector<int>> result = sol.matrixBlockSum(mat, k);

    // Output the result
    cout << "The resulting matrix is:" << endl;
    for (const auto& row : result) {
        for (const auto& val : row) {
            cout << val << " ";
        }
        cout << endl;
    }

    return 0;
}

File: Archita_Banerjee/Assignment-05-Prefix-Suffix-Sum/MatrixBlockSum.java
public class MatrixBlockSum {
    public int[][] matrixBlockSum(int[][] mat, int k) {
        int rows = mat.length;
        int cols = rows == 0 ? 0 : mat[0].length;
        int[][] answer = new int[rows][cols];

        for (int row = 1; row < rows; row++) {
            mat[row][0] += mat[row - 1][0];
        }

        for (int col = 1; col < cols; col++) {
            mat[0][col] += mat[0][col - 1];
        }

        for (int row = 1; row < rows; row++) {
            for (int col = 1; col < cols; col++) {
                mat[row][col] += mat[row - 1][col] + mat[row][col - 1] - mat[row - 1][col - 1];
            }
        }

        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                int r1 = Math.max (0, row - k);
                int c1 = Math.max (0, col - k);
                int r2 = Math.min (rows - 1, row + k);
                int c2 = Math.min (cols - 1, col + k);
                answer[row][col] = mat[r2][c2];
                answer[row][col] -= r1 == 0 ? 0 : mat[r1 - 1][c2];
                answer[row][col] -= c1 == 0 ? 0 : mat[r2][c1 - 1];
                answer[row][col] += r1 == 0 || c1 == 0 ? 0 : mat[r1 - 1][c1 - 1];
            }
        }

        return answer;

    }
}
